File: IE_KNIN_CPP/interpolation.cpp
//
// Created by lisboa on 19/04/2015.
//

#include "interpolation.h"

struct Pdy
{
    std::vector<long double> Y;
    unsigned long size;
};

std::vector<long double> interpolation_mount(struct Point* Pares)
{

    struct Pdy *pdy;

    unsigned long N;
    unsigned long i, j;

    long double temp1;
    long double temp2;

    std::vector<long double> coefficients;

    N = Pares->X.size();
    pdy = new struct Pdy[N];
    pdy[0].Y.reserve(Pares->Y.size());
    pdy[0].Y = Pares->Y;
    pdy[0].size = Pares->Y.size();

    for (i = 1, j = N - 1; i < N; i++, j--)
    {
        pdy[i].Y.reserve(j + 1);
        pdy[i].size = j;
    }

    for (i = 1; i < N; i++)
    {
        for (j = 0; j < pdy[i].size; j++)
        {
            /*
            Formula geral :
            P(i)->Y(j) = ( P(i-1)->Y(j+1) - P(i-1)->Y(j)) / ( X(i+j) - X(j) )
            */
            temp1 = pdy[i - 1].Y.at(j + 1) - pdy[i - 1].Y.at(j);
            temp2 = Pares->X.at(i + j) - Pares->X.at(j);
            pdy[i].Y.push_back(temp1 / temp2);
        }
        coefficients.push_back(pdy[i-1].Y[0]);
    }
    coefficients.push_back(pdy[i-1].Y[0]);
    return coefficients;
}

File: IE_KNIN_JAVAFX/src/sample/IPKnin/Interpolation.java
package sample.IPKnin;

import java.util.ArrayList;

/**
 * Created by lisboa on 03/04/2015.
 */

public class Interpolation {

    ArrayList<Double> X = null;
    ArrayList<Double> Y = null;

    public Interpolation(ArrayList<Double> x, ArrayList<Double> y) {
        this.X = x;
        this.Y = y;
    }

    public double[] mount()
    {
        int N = this.X.size();
        double[] degrees = new double[N];
        double temp1,temp2;

        Pdy[] pdy = new Pdy[N];
        pdy[0] = new Pdy();
        pdy[0].y = this.Y;

        for (int i = 1, j = N - 1; i < N; i++, j--)
        {
            pdy[i] = new Pdy();
            pdy[i].setsize(j);
        }
        for(int i = 1 ; i < N; i++)
            for(int j = 0 ; j < pdy[i].getsize(); j++)
            {
                /*
                Formula geral :
                P(i)->Y(j) = ( P(i-1)->Y(j+1) - P(i-1)->Y(j)) / ( X(i+j) - X(j) )
                */
                temp1 = pdy[i - 1].y.get(j + 1) - pdy[i - 1].y.get(j);
                temp2 = this.X.get(i + j) - this.X.get(j);
                pdy[i].y.add(temp1 / temp2);
            }
        for(int i = 0 ; i < N ; i++)
        {
            degrees[i] = pdy[i].y.get(0);
        }
        return degrees;
    }

}
